sisu-osgi/sisu-osgi-connect/src/main/java/org/eclipse/sisu/osgi/connect/PlexusFrameworkConnectServiceFactory.java

```diff
@@ -102,22 +100,26 @@ protected String getName(ClassLoader classLoader) {
 	 * @return get (or creates) the Framework that is made of the given classloader
 	 * @throws BundleException if creation of the framework failed
 	 */
-	synchronized PlexusConnectFramework getFramework(ClassRealm realm) throws BundleException {
-		PlexusConnectFramework cachedFramework = frameworkMap.get(realm);
-		if (cachedFramework != null) {
-			return cachedFramework;
-		}
+	PlexusConnectFramework getFramework(ClassRealm realm) {
+		return frameworkMap.computeIfAbsent(realm, r -> {
```

Comment:
I don't think that computeIfabset brings any value here beside that we are forced to wrap exceptions in another exception...

Reply:
In combination with using a ConcurrentHashMap it has the advantage that the method itself does not have to be synchronized and therefore a higher level of concurrency (theoretically) can be achieved.
And because the exception was wrapped into another exception in the calling method anyways we don't loose anything in this regard. The wrapping now just happens earlier.

Reply:
I think the key point here is "theoretically" ;-)

Given that the startup of a framework is many decades slower than any synchronization overhead We should not optimize for no good reason here. And just because the caller (might) wrap the exception we should not wrap it here as it makes debugging and tracing harder for no gain.

Reply:
> I think the key point here is "theoretically" ;-)
> 
> Given that the startup of a framework is many decades slower than any synchronization overhead We should not optimize for no good reason here. And just because the caller (might) wrap the exception we should not wrap it here as it makes debugging and tracing harder for no gain.

It was not about the synchronization overhead itself, it was about that in a parallel build, again "theoretically" :D , multiple framework instances could be launched in parallel if different plug-ins request that.
But if you think this is too theoreticall and would better be done with a 'old'-style get-in-case-of-null-compute-and-put I can revert that.

Reply:
I have done it here for different reasons:

1. checked exceptions
2. easier to debug
3. better lock the object (e.g. for maybe other methods as well)
4. code might "call back" while the init state and with `frameworkMap.put(realm, connectFramework);` I can control when this framework will become visible (in this case before `start` was called), while with a concurrentmap is is only visible after the method returns.

so yes I think we better revert this without a strong usecase / analysis, beside that I always use concurrent map if possible :-)

Reply:
>     4. code might "call back" while the init state and with `frameworkMap.put(realm, connectFramework);` I can control when this framework will become visible (in this case before `start` was called), while with a concurrentmap is is only visible after the method returns.

I wonder why one wants to see/use the framework instance before it is completely set up?
Nevertheless I reverted that part.

Reply:
There are two cases:

1. Code using FrameworkUtil.getBundle(class) that can run as we start the framework
2. EquinoxFrameworklisteners calling methods

It might be possible to call start in a lazy way to prevent this but I think we don't need to complicate this even more :-)